lib/SILOptimizer/Mandatory/DiagnoseVarUsage.cpp

```diff
@@ -0,0 +1,1617 @@
+    if (!InitialReference) {
+      InitialReference = getInitialReference();
+    }
+  }
+};
+
+//===----------------------------------------------------------------------===//
+//                         MARK: Diagostics Generator
+//===----------------------------------------------------------------------===//
+
+class UsageDiagnosticsGenerator {
+  DiagnosticEngine &diags;
+  
+public:
+  
+  UsageDiagnosticsGenerator(ASTContext &Context) : diags(Context.Diags) {}
+  
+  void diagnoseDuplicateReference(VarDecl *VD, VarDecl *IVD) {
+    diags.diagnoseWithNotes(
+      diags.diagnose(VD->getLoc(),       
+                     diag::immutable_value_duplicate_reference,
+                     VD->getName(), IVD->getName()),
+                     [&]() {
+      // Emit a note to swap to mark the initial reference
+      diags.diagnose(IVD->getLoc(), diag::value_initial_reference);
+    
+    });
+  }
+  
+  void diagnoseUnreadUnmodified(VarDecl *var, bool DidDeferInit) {
+    
+    // The let is assigned and initialized in seperate lines
+    //    let a: Int
+    //    a = 1
+    if (var->isLet() && DidDeferInit) {
+      diags.diagnose(var->getLoc(),
+                     diag::immutable_value_never_used_but_assigned,
+                     var->getName());
+      return;
+    }
+    
+    // If the source of the VarDecl is a trivial PatternBinding with only a
+    // single binding, rewrite the whole thing into an assignment.
+    //    let x = foo()
+    //  ->
+    //    _ = foo()
+    if (auto *pbd = var->getParentPatternBinding()) {
+      if (pbd->getSingleVar() == var && pbd->getInit(0) != nullptr &&
+          !isa<TypedPattern>(pbd->getPattern(0))) {
+        unsigned varKind = var->isLet();
+        SourceRange replaceRange(pbd->getStartLoc(),
+                                 pbd->getPattern(0)->getEndLoc());
+        diags.diagnose(var->getLoc(), diag::pbd_never_used,
+                       var->getName(), varKind)
+        .fixItReplace(replaceRange, "_");
+        return;
+      }
+    }
+    
+    // If the variable is defined in a pattern in an if/while/guard statement,
+    // see if we can produce a tuned fixit.  When we have something like:
+    //
+    //    if let x = <expr> {
+    //
+    // we prefer to rewrite it to:
+    //
+    //    if <expr> != nil {
+    if (auto ps = var->getParentPatternStmt()) {
+      if (auto SC = dyn_cast<LabeledConditionalStmt>(ps)) {
+        // We only handle the "if let" case right now, since it is vastly the
+        // most common situation that people run into.
+        if (SC->getCond().size() == 1) {
+          auto pattern = SC->getCond()[0].getPattern();
+          if (auto OSP = dyn_cast<OptionalSomePattern>(pattern)) {
+            if (auto LP = dyn_cast<BindingPattern>(OSP->getSubPattern())) {
+              if (isa<NamedPattern>(LP->getSubPattern())) {
+                auto initExpr = SC->getCond()[0].getInitializer();
+                if (initExpr->getStartLoc().isValid()) {
```

Comment:
The deep nesting here makes things a bit hard to read (and there is further nesting later). Perhaps separate it out into a function and use early exits?

Reply:
Okay, I will take a look at reorganizing this.

Reply:
I expanded this out into another function using early exits.